MCLOUD-6375: MC-34668: Introduce environment variable to store a list of quality p…

viktym opened the pull request:
https://jira.corp.magento.com/browse/MCLOUD-6375: 

In scope of the story [Magento Quality Patches](https://jira.corp.magento.com/browse/MC-34580) added environment variable QUALITY_PATCHES. By default, it's empty and doesn't affect the deployment process. We need it in develop branch to proceed with [adopting CI builds](https://jira.corp.magento.com/browse/MC-34666) for Magento Quality Patches 

### Description
- Added environment variable QUALITY_PATCHES
- Updated patch manager to use QUALITY_PATCHES variable
- Implemented copying patch.log file from build phase to {magento_root}/var/log directory

### Testing scenario
- Check if your GitHub token has access to https://github.com/magento-mpi/magento-cloud-patches.git, https://github.com/magento/quality-patches.git, https://github.com/magento-mpi/ece-tools.git
- Add to composer.json 

```
 "repositories": {
        "repo": {
            "type": "composer",
            "url": "https://repo.magento.com"
        },
        "mcp": {
            "type": "vcs",
            "url": "https://github.com/magento-mpi/magento-cloud-patches.git"
        },
        "mqp": {
            "type": "vcs",
            "url": "https://github.com/magento/quality-patches.git"
        },
        "ece": {
            "type": "vcs",
            "url": "https://github.com/magento-mpi/ece-tools.git"
        }
    },
    "require": {
        "magento/magento-cloud-patches": "dev-MC-34580 as 1.0.99",
        "magento/quality-patches": "dev-master as 1.0.999",
        "magento/ece-tools": "dev-MC-34668 as 2002.1.99"
    },
```

- Add to .magento.env.yaml
```
 stage:
    build:
        QUALITY_PATCHES:
            - MCTEST-1001
```

- Install Magento
- Check /var/log/patch.log file contains 
```
INFO: Patch MCTEST-1001 has been applied {"file":"/app/vendor/magento/quality-patches/src/../patches/MCTEST-1001__TEST_changes__2.3.5_ce.patch"}
INFO: Patch MCTEST-1001 has been applied {"file":"/app/vendor/magento/quality-patches/src/../patches/MCTEST-1001__TEST_changes__2.3.5_ee.patch"}
```
- Verify that patches https://github.com/magento/quality-patches/blob/master/patches/MCTEST-1001__TEST_changes__2.3.5_ce.patch and https://github.com/magento/quality-patches/blob/master/patches/MCTEST-1001__TEST_changes__2.3.5_ee.patch were applied (check sources in vendor folder) 

### Release notes

Added the `QUALITY_PATCHES` environment variable to specify one or more patches to apply during the deployment process. Currently, this variable is for Magento internal use only.

### Associated documentation updates



### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] Pull request introduces user-facing changes and includes meaningful updates for any required release notes and documentation changes
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

shiftedreality commented on config/schema.yaml:
```diff
@@ -643,6 +643,18 @@ variables:
+    examples:
+      - stage:
+          build:
+            QUALITY_PATCHES: 'MC-31387 MDVA-4567 MC-45634'
```
I would suggest using a native array structure in YAML

viktym replied:
Yeah, I thought about it, but CLI command suppose to take a wildcard as well
`./vendor/bin/ece-patches apply --patches="*"` to apply all optional patches. It was a decision to not advertise the possibility of using wildcard since we don't want to encourage customers to apply a big amount of patches. But we need it internally to run CI build with all patches applied (should guaranty that it's possible to apply all patches without conflicts). To be consistent with 'patches'  command option type I made it a string

viktym replied:
I reconsidered how the list of patches should pass in CLI command, changed from option to argument
`./vendor/bin/ece-patches apply --patches="MC-1111 MC-2222"` => `./vendor/bin/ece-patches apply MC-1111 MC-2222`. So QUALITY_PATCHES variable updated with an array type. Now it looks better. Thanks.

YPyltiai replied:
@viktym maybe we could name the argument quality-patches instead of just patches to be more explicit? There are chances we can do current required patches optional later via similar argument, so would be good to adjust for that kind of situation.

viktym replied:
@YPyltiai right now there is no option name, the list of patches passes as argument of the command:
`./vendor/bin/ece-patches apply MC-1111 MC-2222 MDVA-4356`
If we need to make some of the patches from magento/magento-cloud-patches optional, we can move them to magento/quality-patches

YPyltiai replied:
@viktym that makes sense. We can proceed as is. Thanks for clarification.

shiftedreality commented on src/Test/Unit/Config/SchemaTest.php:
```diff
@@ -77,6 +77,7 @@ public function testGetDefaultsForBuild(): void
                 BuildInterface::VAR_SCD_MAX_EXEC_TIME => null,
                 BuildInterface::VAR_ERROR_REPORT_DIR_NESTING_LEVEL => 1,
                 BuildInterface::VAR_SCD_USE_BALER => false,
+                BuildInterface::VAR_QUALITY_PATCHES => ''
```
I believe you will have to bump the `magento-cloud-patches` dependency in `composer.json`

BaDos replied:
It is for https://github.com/magento/quality-patches

billygilbert commented on src/Test/Unit/Step/Build/ApplyPatchesTest.php:
```diff
@@ -31,23 +32,34 @@ class ApplyPatchesTest extends TestCase
+            ->willReturn($qualityPatches);
         $this->managerMock->expects($this->once())
-            ->method('apply');
+            ->method('apply')->with($qualityPatches);
```
Move to new line

BaDos replied:
@billygilbert I will help with it.
Also, I extend our FT for `magento-cloud-patches` to check that patches do not have any conflicts from both repos.

viktym replied:
Fixed

BaDos replied:
My comment is not for code, it is an answer for the general comment above.

billygilbert reviewed (changes requested):
One change requested.  Please ensure that you have tasks on your backlog to cover updating Cloud DevDocs as well as proper cloud release notes.

viktym committed:
MC-34668: Introduce environment variable to store a list of quality patches for Cloud customers

- Added environment variable QUALITY_PATCHES
- Updated patch manager to use QUALITY_PATCHES variable

billygilbert reviewed (changes requested):
@viktym Static tests failed due to line length, please resolve:
src/Test/Unit/Patch/ManagerTest.php
line 111

viktym commented:
> @viktym Static tests failed due to line length, please resolve:
> src/Test/Unit/Patch/ManagerTest.php
> line 111

@billygilbert  Fixed

viktym committed:
Merge remote-tracking branch 'magento/develop' into MC-34668

oshmyheliuk commented on src/Patch/Manager.php:
```diff
@@ -51,19 +51,26 @@ public function __construct(
         $command = 'php ./vendor/bin/ece-patches apply';
 
+        if ($qualityPatches) {
+            $command .= sprintf(" '%s'", implode("' '", $qualityPatches));
```
To use this functionality we must ensure that client has the latest `magento/magento-cloud-patches` version.
I think, before next release of `ece-tools` constraints in composer.json should be updated to `"magento/magento-cloud-patches": "^1.0.6"`

viktym committed:
MC-34668: Introduce environment variable to store a list of quality patches for Cloud customers

- Added step to restore patch.log file from build phase

viktym committed:
Merge remote-tracking branch 'magento/develop' into MC-34668

# Conflicts:
#	src/Filesystem/FileList.php

viktym committed:
MC-34668: Introduce environment variable to store a list of quality patches for Cloud customers

- Added step to restore patch.log file from build phase

viktym committed:
Merge remote-tracking branch 'magento/develop' into MC-34668

viktym committed:
Merge remote-tracking branch 'magento/develop' into MC-34668

viktym committed:
Merge remote-tracking branch 'magento/develop' into MC-34668

mveeramneni commented:
QA Approved

ghost commented:
Hi @viktym, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+ece-tools+750), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
